presto-cache/src/main/java/com/facebook/presto/cache/alluxio/AlluxioCachingFileSystem.java

```diff
@@ -0,0 +1,247 @@
+
+    @Override
+    public Path getWorkingDirectory()
+    {
+        return cachingFileSystem.getWorkingDirectory();
+    }
+
+    @Override
+    public boolean mkdirs(Path path, FsPermission permission)
+            throws IOException
+    {
+        return cachingFileSystem.mkdirs(path, permission);
+    }
+
+    @Override
+    public FileStatus getFileStatus(Path path)
+            throws IOException
+    {
+        return cachingFileSystem.getFileStatus(path);
+    }
+
+    @Override
+    public FSDataInputStream openFile(Path path, HiveFileContext hiveFileContext)
+            throws Exception
+    {
+        if (hiveFileContext.isCacheable()) {
+            return cachingFileSystem.openFile(path, hiveFileContext);
+        }
+        return fileSystem.openFile(path, hiveFileContext);
+    }
+
+    private class AlluxioCachingFileSystemInternal
+            extends AbstractFileSystem
+    {
+        // The filesystem to query on cache miss
+        private final ExtendedFileSystem fileSystem;
+        private final URI uri;
+
+        AlluxioCachingFileSystemInternal(Configuration configuration, URI uri, ExtendedFileSystem fileSystem)
+        {
+            this.fileSystem = requireNonNull(fileSystem, "filesystem is null");
+            this.uri = requireNonNull(uri, "uri is null");
+        }
+
+        @Override
+        public synchronized void initialize(URI uri, Configuration configuration)
+                throws IOException
+        {
+            // Set statistics
+            setConf(configuration);
+            statistics = getStatistics(uri.getScheme(), getClass());
+
+            // Take the URI properties, hadoop configuration, and given Alluxio configuration and merge
+            // all three into a single object.
+            Map<String, Object> uriConfProperties = getConfigurationFromUri(uri);
+            AlluxioProperties alluxioProperties = ConfigurationUtils.defaults();
+            InstancedConfiguration newConfiguration = HadoopConfigurationUtils.mergeHadoopConfiguration(configuration, alluxioProperties);
+            // Connection details in the URI has the highest priority
+            newConfiguration.merge(uriConfProperties, Source.RUNTIME);
+            mAlluxioConf = newConfiguration;
+
+            // Handle metrics
+            Properties metricsProperties = new Properties();
+            for (Map.Entry<String, String> entry : configuration) {
+                metricsProperties.setProperty(entry.getKey(), entry.getValue());
+            }
+            MetricsSystem.startSinksFromConfig(new MetricsConfig(metricsProperties));
+            mFileSystem = new LocalCacheFileSystem(new AlluxioCachingClientFileSystem(fileSystem, mAlluxioConf), mAlluxioConf);
+            super.initialize(uri, configuration);
+        }
+
+        public FSDataInputStream openFile(Path path, HiveFileContext hiveFileContext)
+                throws Exception
+        {
+            return fileSystem.openFile(path, hiveFileContext);
```

Comment:
This is not going to work. `openFile` directly delegates to the underlying file system. The input stream does not use cache at all

Reply:
Yes! Can you please use your cacheManager interface to pass the object around. As I commented above use URIStatus.

Reply:
sorry, I think this is due to some merge conflict btw my change and @bf8086 's

However, do you want alluxio local cache to support Presto  `open(Path, HiveFileContext)` API?
That is not doable without have a new release of Alluxio client

Reply:
> Yes! Can you please use your cacheManager interface to pass the object around. As I commented above use URIStatus.
Let me try to play with it

Reply:
We always wanted to support it in the open source. I am not sure why do you say that. We created a mechanism via URIStatus. I can fix the code in the morning if you are not sure.

Reply:
yes I remember you and Calvin solved this issue before. Sorry I don't have the details so feel free to modify my current PR